From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Tue, 26 May 2020 21:39:55 +0200
Subject: [PATCH] Add clear reputation API


diff --git a/src/main/java/org/bukkit/entity/Villager.java b/src/main/java/org/bukkit/entity/Villager.java
index e0a3285b8ca97f6d56e5548d1f95efc1cc7cbd3c..192ba39a3c4eaf31e0cec962fbbdd45b964958cb 100644
--- a/src/main/java/org/bukkit/entity/Villager.java
+++ b/src/main/java/org/bukkit/entity/Villager.java
@@ -266,5 +266,12 @@ public interface Villager extends AbstractVillager {
      * for all players mapped by their {@link UUID unique IDs}.
      */
     public void setReputations(@NotNull Map<UUID, com.destroystokyo.paper.entity.villager.Reputation> reputations);
+
+    /**
+     * Clear all reputations associated with this villager.
+     *
+     * @author Papercut
+     */
+    public void clearReputations();
     // Paper end
 }
