From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nahuel <nahueldolores@hotmail.com>
Date: Thu, 28 May 2020 21:42:36 -0300
Subject: [PATCH] Make StructureGenerateEvent fire async if it's configured to


diff --git a/src/main/java/com/destroystokyo/paper/event/world/StructureGenerateEvent.java b/src/main/java/com/destroystokyo/paper/event/world/StructureGenerateEvent.java
index e00d5a598ca2eb1fb14a41712f155fb506546498..3f3d9416e4a335045c71eecca2298892dc146b6d 100644
--- a/src/main/java/com/destroystokyo/paper/event/world/StructureGenerateEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/world/StructureGenerateEvent.java
@@ -18,8 +18,8 @@ public class StructureGenerateEvent extends Event implements Cancellable {
 
     private boolean cancel = false;
 
-    public StructureGenerateEvent(@NotNull World world, @NotNull StructureType structureType, int chunkX, int chunkZ) {
-        super(true);
+    public StructureGenerateEvent(boolean async, @NotNull World world, @NotNull StructureType structureType, int chunkX, int chunkZ) {
+        super(async);
         this.world = Preconditions.checkNotNull(world, "world");
         this.structureType = Preconditions.checkNotNull(structureType, "structureType");
         this.chunkX = chunkX;
