From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nahuel <nahueldolores@hotmail.com>
Date: Tue, 28 Apr 2020 13:52:06 -0300
Subject: [PATCH] Add StructureGenerateEvent

---
 .../net/minecraft/server/ChunkGenerator.java  | 23 +++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/src/main/java/net/minecraft/server/ChunkGenerator.java b/src/main/java/net/minecraft/server/ChunkGenerator.java
index ee6726e5c664115417c6397aa153315c8eb835a9..92cc25423d2e9fdcc2274b7e303e282e3ecca185 100644
--- a/src/main/java/net/minecraft/server/ChunkGenerator.java
+++ b/src/main/java/net/minecraft/server/ChunkGenerator.java
@@ -218,6 +218,7 @@ public abstract class ChunkGenerator {
             // CraftBukkit start
             if (structurefeature.b == StructureGenerator.STRONGHOLD) {
                 synchronized (structurefeature) {
+                    // TODO(Papercut): add StructureGenerateEvent
                     this.a(structurefeature, structuremanager, ichunkaccess, definedstructuremanager, i, chunkcoordintpair, biomebase);
                 }
             } else
@@ -314,6 +315,28 @@ public abstract class ChunkGenerator {
     // Spigot end
 
     public void storeStructures(GeneratorAccess generatoraccess, StructureManager structuremanager, IChunkAccess ichunkaccess) {
+//=======
+//    // Papercut start
+//    private StructureStart getStructureStart(BiomeManager biomemanager, ChunkGenerator<?> chunkgenerator, DefinedStructureManager definedstructuremanager, StructureGenerator<?> structuregenerator, int i, SeededRandom seededrandom, ChunkCoordIntPair chunkcoordintpair, StructureStart structurestart1, BiomeBase biomebase) {
+//        StructureGenerateEvent event = new StructureGenerateEvent(
+//                PaperConfig.asyncChunks && !getWorld().getServer().isPrimaryThread(),
+//                getWorld().getWorld(),
+//                StructureType.getStructureTypes().get(structuregenerator.b().toLowerCase()),
+//                chunkcoordintpair.x,
+//                chunkcoordintpair.z
+//        );
+//        if (structuregenerator.a(biomemanager, chunkgenerator, seededrandom, chunkcoordintpair.x, chunkcoordintpair.z, biomebase) && event.callEvent()) {
+//            StructureStart structurestart2 = structuregenerator.a().create(structuregenerator, chunkcoordintpair.x, chunkcoordintpair.z, StructureBoundingBox.a(), i, chunkgenerator.getSeed());
+//
+//            structurestart2.a(this, definedstructuremanager, chunkcoordintpair.x, chunkcoordintpair.z, biomebase);
+//            structurestart1 = structurestart2.e() ? structurestart2 : StructureStart.a;
+//        }
+//        return structurestart1;
+//    }
+//    // Papercut end
+//
+//    public void storeStructures(GeneratorAccess generatoraccess, IChunkAccess ichunkaccess) {
+//>>>>>>> Add StructureGenerateEvent
         boolean flag = true;
         int i = ichunkaccess.getPos().x;
         int j = ichunkaccess.getPos().z;
