From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Tue, 26 May 2020 21:23:51 +0200
Subject: [PATCH] Ban EntityTrackerFixer


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 34775206287ca1221b1dc58793d49e8c5da47f91..4978cb5659ecb4649c8dd8e6db8c271b17fbefc3 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -383,6 +383,12 @@ public final class CraftServer implements Server {
 
         for (Plugin plugin : plugins) {
             if ((!plugin.isEnabled()) && (plugin.getDescription().getLoad() == type)) {
+                // Papercut start
+                if ("net.minemora.entitytrackerfixer.EntityTrackerFixer".equals(plugin.getDescription().getMain())) {
+                    plugin.getLogger().warning("EntityTrackerFixer is unsupported on Papercut.");
+                    return;
+                }
+                // Papercut end
                 enablePlugin(plugin);
             }
         }
