From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Tue, 26 May 2020 21:40:01 +0200
Subject: [PATCH] Add clear reputation API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
index 19409c7a25c48f3f72f7e0b6306b81aa90b38f7c..d24a892c498d7ee58741c9358748a117f01d8a8d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
@@ -163,5 +163,10 @@ public class CraftVillager extends CraftAbstractVillager implements Villager {
             setReputation(entry.getKey(), entry.getValue());
         }
     }
+
+    @Override
+    public void clearReputations() {
+        getHandle().getReputation().getReputations().clear();
+    }
     // Paper end
 }
