From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Tue, 26 May 2020 22:20:23 +0200
Subject: [PATCH] Apply advancements async


diff --git a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
index c41e1384724ab150f43dc43fe2a453c9b1262e48..3d7db96b9a25dcc184850a09eeb9e3064c40b9c7 100644
--- a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
+++ b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
@@ -56,6 +56,7 @@ public class AdvancementDataPlayer {
         this.g();
     }
 
+    public void setPlayer(EntityPlayer entityplayer) { a(entityplayer); } // Paper - OBFHELPER
     public void a(EntityPlayer entityplayer) {
         this.player = entityplayer;
     }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index d60f659b368500e3a8c3305f99e60ffc643e2fbd..b54c7bad28cbe9408a4e261eccbb5a8417214203 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -47,7 +47,6 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public final MinecraftServer server;
     public final PlayerInteractManager playerInteractManager;
     public final Deque<Integer> removeQueue = new ArrayDeque<>(); // Paper
-    private final AdvancementDataPlayer advancementDataPlayer;
     private final ServerStatisticManager serverStatisticManager;
     private float lastHealthScored = Float.MIN_VALUE;
     private int lastFoodScored = Integer.MIN_VALUE;
@@ -120,7 +119,6 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.server = minecraftserver;
         this.recipeBook = new RecipeBookServer(minecraftserver.getCraftingManager());
         this.serverStatisticManager = minecraftserver.getPlayerList().getStatisticManager(this);
-        this.advancementDataPlayer = minecraftserver.getPlayerList().f(this);
         this.H = 1.0F;
         //this.a(worldserver); // Paper - don't move to spawn on login, only first join
 
@@ -433,7 +431,8 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             CriterionTriggers.u.a(this, this.co, this.ticksLived - this.cp);
         }
 
-        this.advancementDataPlayer.b(this);
+        if (isAdvancementDataLoaded()) // Paper - async advancement loading
+            getAdvancementData().b(this); // Paper - async advancement loading
     }
 
     public void playerTick() {
@@ -1758,8 +1757,9 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     }
 
     public AdvancementDataPlayer getAdvancementData() {
-        return this.advancementDataPlayer;
+        return this.playerConnection.advancementDataPlayer.join();
     }
+    public boolean isAdvancementDataLoaded() { return this.playerConnection != null && this.playerConnection.advancementDataPlayer != null && this.playerConnection.advancementDataPlayer.isDone(); } // Paper - async advancements loading
 
     // CraftBukkit start
     public void a(WorldServer worldserver, double d0, double d1, double d2, float f, float f1) {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index b25b3b48165e5fef4db99c2838de21d4eca502f4..433efa13ab758080967ee325e2b7df65d27cd95d 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -106,6 +106,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
     private int processedMovePackets;
     private static final int MAX_SIGN_LINE_LENGTH = Integer.getInteger("Paper.maxSignLength", 80);
     private static final long KEEPALIVE_LIMIT = Long.getLong("paper.playerconnection.keepalive", 30) * 1000; // Paper - provide property to set keepalive limit
+    java.util.concurrent.CompletableFuture<AdvancementDataPlayer> advancementDataPlayer = null; // Paper - async player advancements loading
 
     public PlayerConnection(MinecraftServer minecraftserver, NetworkManager networkmanager, EntityPlayer entityplayer) {
         this.minecraftServer = minecraftserver;
@@ -537,8 +538,8 @@ public class PlayerConnection implements PacketListenerPlayIn {
             MinecraftKey minecraftkey = packetplayinadvancements.d();
             Advancement advancement = this.minecraftServer.getAdvancementData().a(minecraftkey);
 
-            if (advancement != null) {
-                this.player.getAdvancementData().a(advancement);
+            if (advancement != null && this.player.isAdvancementDataLoaded()) { // Paper - async advancements
+                player.getAdvancementData().a(advancement); // Paper
             }
         }
 
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index ada082e67ac8a3e79bab5b360f09c9402f683b86..a78e50c3114aaef9ac43efd0a5de31c88bde5721 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -16,6 +16,7 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
 import javax.annotation.Nullable;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -448,10 +449,8 @@ public abstract class PlayerList {
             serverstatisticmanager.a();
         }
 
-        AdvancementDataPlayer advancementdataplayer = (AdvancementDataPlayer) entityplayer.getAdvancementData(); // CraftBukkit
-
-        if (advancementdataplayer != null) {
-            advancementdataplayer.c();
+        if (entityplayer.isAdvancementDataLoaded()) { // Paper
+            entityplayer.getAdvancementData().c(); // Paper
         }
 
     }
@@ -503,7 +502,7 @@ public abstract class PlayerList {
 
         entityplayer.decouple();
         worldserver.removePlayer(entityplayer);
-        entityplayer.getAdvancementData().a();
+        if (entityplayer.isAdvancementDataLoaded()) entityplayer.getAdvancementData().a(); // Paper
         this.players.remove(entityplayer);
         this.playersByName.remove(entityplayer.getName().toLowerCase(java.util.Locale.ROOT)); // Spigot
         this.server.getBossBattleCustomData().b(entityplayer);
@@ -622,6 +621,14 @@ public abstract class PlayerList {
             loginlistener.disconnect(event.getKickMessage());
             return null;
         }
+        // Paper start - async advancements loading
+        // At this point, the player should be allowed in.
+        // In almost all cases, loading advancements is QUICK, so doing this async should have little impact.
+        // Due to that same reason, starting it this late should also be okay.
+        server.execute(() ->
+            entity.playerConnection.advancementDataPlayer = server.getPlayerList().applyPlayerAdvancementsAsync(entity)
+        );
+        // Paper end
         return entity;
     }
 
@@ -1260,9 +1267,40 @@ public abstract class PlayerList {
         return serverstatisticmanager;
     }
 
+    // Paper start - Apply advancements async
+    public java.util.concurrent.CompletableFuture<AdvancementDataPlayer> applyPlayerAdvancementsAsync(EntityPlayer entityplayer) {
+        if (!us.minevict.papercut.PapercutConfig.applyAdvancementsAsync) {
+            return CompletableFuture.completedFuture(applyPlayerAdvancementsSync(entityplayer));
+        }
+        if (entityplayer.isAdvancementDataLoaded()) {
+            // We have nothing to load, so no reason to drag on the future for long.
+            AdvancementDataPlayer advancementdataplayer = entityplayer.getAdvancementData();
+            java.util.concurrent.CompletableFuture<AdvancementDataPlayer> future = java.util.concurrent.CompletableFuture.completedFuture(advancementdataplayer);
+            advancementdataplayer.setPlayer(entityplayer);
+            return future;
+        }
+
+        // This time we were not so lucky, and advancements have to be read from file.
+        // That is slow...
+        UUID uuid = entityplayer.getUniqueID();
+        File file = new File(this.server.getWorldServer(DimensionManager.OVERWORLD).getDataManager().getDirectory(), "advancements");
+        File file1 = new File(file, uuid + ".json");
+        java.util.concurrent.CompletableFuture<AdvancementDataPlayer> future = new java.util.concurrent.CompletableFuture<AdvancementDataPlayer>();
+        this.server.execute(() -> {
+            AdvancementDataPlayer data = new AdvancementDataPlayer(this.server, file1, entityplayer);
+            data.setPlayer(entityplayer);
+            // this.p.put(uuid, advancementdataplayer); // CraftBukkit
+            future.complete(data);
+        });
+
+        return future;
+    }
+    // Paper end
+
+    public AdvancementDataPlayer applyPlayerAdvancementsSync(EntityPlayer entityplayer) { return f(entityplayer); } // Papercut - OBFHELPER
     public AdvancementDataPlayer f(EntityPlayer entityplayer) {
         UUID uuid = entityplayer.getUniqueID();
-        AdvancementDataPlayer advancementdataplayer = (AdvancementDataPlayer) entityplayer.getAdvancementData(); // CraftBukkit
+        AdvancementDataPlayer advancementdataplayer = entityplayer.isAdvancementDataLoaded() ? entityplayer.getAdvancementData() : null; // CraftBukkit // Paper
 
         if (advancementdataplayer == null) {
             File file = new File(this.server.getWorldServer(DimensionManager.OVERWORLD).getDataManager().getDirectory(), "advancements");
@@ -1315,8 +1353,13 @@ public abstract class PlayerList {
         }*/
 
         for (EntityPlayer player : players) {
-            player.getAdvancementData().b();
-            player.getAdvancementData().b(player); // CraftBukkit - trigger immediate flush of advancements
+            // Paper start - null check on AdvancementDataPlayer
+            AdvancementDataPlayer advancementDataPlayer = player.getAdvancementData();
+            if (advancementDataPlayer != null) {
+                advancementDataPlayer.b();
+                advancementDataPlayer.b(player); // CraftBukkit - trigger immediate flush of advancements // Paper
+            }
+            // Paper end
         }
         // CraftBukkit end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java
index 60fc5aff80697cb4c85080350542c0f46056f74a..304362b4c3f2edc87478201cc73cb857e43f567b 100644
--- a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java
+++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java
@@ -33,11 +33,13 @@ public class CraftAdvancementProgress implements AdvancementProgress {
 
     @Override
     public boolean awardCriteria(String criteria) {
+        if (playerData == null) return false; // Paper - null check
         return playerData.grantCriteria(advancement.getHandle(), criteria);
     }
 
     @Override
     public boolean revokeCriteria(String criteria) {
+        if (playerData == null) return false; // Paper - null check
         return playerData.revokeCritera(advancement.getHandle(), criteria);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 787995e81388edcde786fa5da719552092d8fad1..25c1eda3a9cef2c8d1c485d903667abd5436f040 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1876,7 +1876,6 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         CraftAdvancement craft = (CraftAdvancement) advancement;
         AdvancementDataPlayer data = getHandle().getAdvancementData();
         AdvancementProgress progress = data.getProgress(craft.getHandle());
-
         return new CraftAdvancementProgress(craft, data, progress);
     }
 
diff --git a/src/main/java/us/minevict/papercut/PapercutConfig.java b/src/main/java/us/minevict/papercut/PapercutConfig.java
index 878d1b6c7d3382203e200bb152c290df40e6725d..473063016bf4b30e40e2c0bd26a8c4555834df85 100644
--- a/src/main/java/us/minevict/papercut/PapercutConfig.java
+++ b/src/main/java/us/minevict/papercut/PapercutConfig.java
@@ -102,6 +102,11 @@ public final class PapercutConfig {
     return PapercutConfig.config.getDouble(path, dfl);
   }
 
+  public static boolean applyAdvancementsAsync = true;
+  private static void applyAdvancementsAsync() {
+    applyAdvancementsAsync = getBoolean("apply-async-advancements", applyAdvancementsAsync);
+  }
+
   public static final class WorldConfig {
     public final String worldName;
     public ConfigurationSection config;
