From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Sun, 28 Jun 2020 14:03:03 +0200
Subject: [PATCH] Kick players before disabling plugins

---
 .../net/minecraft/server/MinecraftServer.java | 22 ++++++++++++++-----
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 18cf21778265507e304c143fbd8c1b93e6249095..f6cb3ea7dda4eb17529097e2257cb2fd516de3df 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -769,6 +769,14 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         // CraftBukkit end
         MinecraftServer.LOGGER.info("Stopping server (Ignore any thread death message you see! - DO NOT REPORT THREAD DEATH TO PAPER)"); // Paper
         MinecraftTimings.stopServer(); // Paper
+        // Papercut start - move up
+        if (this.playerList != null) {
+            MinecraftServer.LOGGER.info("Saving players");
+            this.playerList.savePlayers();
+            this.playerList.shutdown(this.isRestarting); // Paper
+            try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
+        }
+        // Papercut end
         // CraftBukkit start
         if (this.server != null) {
             this.server.disablePlugins();
@@ -779,12 +787,14 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
             this.getServerConnection().b();
         }
 
-        if (this.playerList != null) {
-            MinecraftServer.LOGGER.info("Saving players");
-            this.playerList.savePlayers();
-            this.playerList.shutdown(this.isRestarting); // Paper
-            try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
-        }
+        // Papercut start - move up
+//        if (this.playerList != null) {
+//            MinecraftServer.LOGGER.info("Saving players");
+//            this.playerList.savePlayers();
+//            this.playerList.shutdown(this.isRestarting); // Paper
+//            try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
+//        }
+        // Papercut end
 
         MinecraftServer.LOGGER.info("Saving worlds");
         Iterator iterator = this.getWorlds().iterator();
