From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Thu, 30 Jul 2020 11:57:24 +0200
Subject: [PATCH] Wither Skulls die after 60 sec

---
 .../java/net/minecraft/server/EntityWitherSkull.java   | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/main/java/net/minecraft/server/EntityWitherSkull.java b/src/main/java/net/minecraft/server/EntityWitherSkull.java
index 72d59adce78943ce47e91301fe17e96ec3579fa6..78a48218bdbc0b2b451ac13887700594125eddfc 100644
--- a/src/main/java/net/minecraft/server/EntityWitherSkull.java
+++ b/src/main/java/net/minecraft/server/EntityWitherSkull.java
@@ -116,4 +116,14 @@ public class EntityWitherSkull extends EntityFireball {
     protected boolean Y_() {
         return false;
     }
+
+    // Papercut start - despawn wither skulls after 60 seconds
+    @Override
+    public void tick() {
+        if (ticksLived >= 60 * 20) {
+            die();
+        }
+        super.tick();
+    }
+    // Papercut end
 }
